### ISSUE TITLE
Add "<function> defined here" notes to more errors

### ISSUE THREAD

**Issue by msullivan** [python/mypy#4773]
mypy produces a nice "<function> defined here" note when calling a function with unexpected kwargs, but in no other situation. We should consider adding the note to *every* function call typechecking failure.

Probably should wait until #4772 is fixed though?

**Comment by MatthieuTPHR** [comment]
Agreed with the comment from @tommyjcarpenter  !

**Comment by udiNur** [comment]
> I find this confusing when using editor linters like Ale, without more info included other than "defined here". There will be a break point to the left of the `def foo` line and it just says "function foo defined here", like, "ok, so??" it doesn't give any additional context of why it is telling me this or what the problem is. Nicer would be "defined here but a caller is calling incorrectly" or something along those lines.
> 
> I came across this issue when googling what on earth this linter note even means.


I completely agree. It's frustrating that the linter highlights the function definition rather than the incorrect calls to that function. On top of that, the user also gets a vague message like "function foo defined here," which doesn't provide any useful context. It would be much more helpful if it specified something like "defined here, but called incorrectly elsewhere." This lack of clarity led me to search for more information online to understand what the linter was trying to convey.

**Comment by tsheaff** [comment]
![CleanShot 2024-12-07 at 21 47 27](https://github.com/user-attachments/assets/1b9c9201-22dd-4352-9ef6-c019bde0111f)

This even confuses an LLM. This is Cursor 3.5 Sonnet's "Fix in Composer" and the LLM is like "doesn't seem like an error to me, it's defined right here" which is, to be fair to Claude, exactly what I thought when I read it the first time too 😆

**Comment by BackSeat** [comment]
Came here to find out why mypy is telling me that my function is defined 'here'. Odd message. Now I understand what it means, but the message shouldn't need me to search to find out what it means.

**Comment by nonamethanks** [comment]
How can this be disabled? It's not really helpful and if anything it's extremely annoying in any IDE that was developed in the last century.

**Comment by TommyDew42** [comment]
Should we just remove the note? It brings mostly confusions to users.

**Comment by JoElfner** [comment]
Yea please, I don't see no reason to keep something which has no use, completely different from typical mypy error messages and only confusing.

**Comment by ilevkivskyi** [comment]
I actually have some counter-proposals. FWIW I myself find this note confusing and noisy. So I would propose to either:
1. Delete this completely.
2. Make this note opt-in, with `--show-func-defs` (then we can show it for other call-related errors)
3. Make this note regularly formatted, for example:
   ```
   main.py:5: error: Unexpected keyword argument "location" for "method" of "Class"
   main.py:5: note: "method" has arguments named "position", "other"
   main.py:5: note: "method" defined in "pkg.mod"
   ```
   again, in this case we can show this for other call-related errors.

I would prefer the option 3. cc @JukkaL
